activesupport/lib/active_support/cache/key.rb

```diff
@@ -0,0 +1,60 @@
+# frozen_string_literal: true
+
+module ActiveSupport
+  module Cache
+    # This class is responsible for coercing input into a valid
+    # key format and cache version for cache stores.
+    #
+    # The goal of this class is to provide a composable way
+    # to build keys. You can put an instance of Key
+    # into another instance of Key and it should produce
+    # the same `cache_key` and `cache_version`.
+    #
+    # Example:
+    #
+    #   k1 = Key.new("foo")
+    #   k2 = Key.new(k1)
+    #   k1.cache_key == k2.cache_key
+    #   # => true
+    #
+    class Key # :nodoc:
+      @cache_object = nil
+
+      def initialize(key, cache_object = self.class.default_cache_object)
+        @key_parts = [key]
+        @cache_key = nil
+        @cache_version = nil
+        @cache_object = cache_object
+      end
+
+      def cache_key
+        @cache_key ||= @cache_object.send(:expanded_key, @key_parts)
+      end
+
+      def cache_version
+        @cache_version ||= @cache_object.send(:expanded_version, @key_parts)
+      end
+
+      def cache_key_with_version
+        @cache_object.send(:retrieve_cache_key, @key_parts)
```

Comment:
Not too keen on the `_object` variable suffix since everything is an object in Ruby, so it doesn't signify much 😄 How about `@cache_store`?